src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualChar.cs

```diff
@@ -3,64 +3,178 @@
+            => new VirtualChar(rune, surrogateChar: default, span);
+
+        /// <summary>
+        /// Creates a new <see cref="VirtualChar"/> from an unpaired high or low surrogate character.  This will throw
+        /// if <paramref name="surrogateChar"/> is not actually a surrogate character. The resultant <see cref="Rune"/>
+        /// value will be <see cref="Rune.ReplacementChar"/>.
+        /// </summary>
+        public static VirtualChar Create(char surrogateChar, TextSpan span)
         {
+            if (!char.IsSurrogate(surrogateChar))
+                throw new ArgumentException(nameof(surrogateChar));
+
+            return new VirtualChar(rune: Rune.ReplacementChar, surrogateChar, span);
+        }
+
+        private VirtualChar(Rune rune, char surrogateChar, TextSpan span)
+        {
+            Contract.ThrowIfFalse(surrogateChar == 0 || rune == Rune.ReplacementChar,
+                "If surrogateChar is provided then rune must be Rune.ReplacementChar");
+
             if (span.IsEmpty)
-            {
                 throw new ArgumentException("Span should not be empty.", nameof(span));
-            }
 
-            Char = @char;
+            Rune = rune;
+            SurrogateChar = surrogateChar;
             Span = span;
         }
 
+        /// <summary>
+        /// Retrieves the scaler value of this character as an <see cref="int"/>.  If this is an unpaired surrogate
+        /// character, this will be the value of that surrogate.  Otherwise, this will be the value of our <see
+        /// cref="Rune"/>.
+        /// </summary>
+        public int Value => SurrogateChar != 0 ? SurrogateChar : Rune.Value;
+
+        #region equality
+
+        public static bool operator ==(VirtualChar char1, VirtualChar char2)
+            => char1.Equals(char2);
+
+        public static bool operator !=(VirtualChar char1, VirtualChar char2)
+            => !(char1 == char2);
+
+        public static bool operator ==(VirtualChar ch1, char ch2)
+            => ch1.Value == ch2;
+
+        public static bool operator !=(VirtualChar ch1, char ch2)
+            => !(ch1 == ch2);
+
         public override bool Equals(object obj)
             => obj is VirtualChar vc && Equals(vc);
 
         public bool Equals(VirtualChar other)
-            => Char == other.Char &&
+            => Rune == other.Rune &&
+               SurrogateChar == other.SurrogateChar &&
                Span == other.Span;
 
         public override int GetHashCode()
         {
-            unchecked
+            var hashCode = 1985253839;
+            hashCode = hashCode * -1521134295 + Rune.GetHashCode();
+            hashCode = hashCode * -1521134295 + SurrogateChar.GetHashCode();
+            hashCode = hashCode * -1521134295 + Span.GetHashCode();
+            return hashCode;
+        }
+
+        #endregion
+
+        #region string operations
+
+        public override string ToString()
+            => Rune.ToString();
```

Comment:
❔ Is this still the operation we want, now that it can handle unpaired surrogates?

Reply:
good catch.